### ISSUE TITLE
"--debug-jvm" broken since JDK 9 when running Gradle inside a container and debugging from the host

### ISSUE THREAD

**Issue by ChristianCiach** [gradle/gradle#7496]
I often debug unit tests run by Gradle. Gradle runs inside a (Docker-) container and I map the port 5005 from the container to the host.

This worked fine with JDK 8, but is broken since JDK 9. (Or so I heared. I actually build with JDK 11 now.)

It is broken because JDK 9 only accepts debug connections originating from the same host. Please see this answer at Stackoverflow: https://stackoverflow.com/a/47772365/1507544

So, when running unit tests with JDK 9 and onwards, instead of these JVM args ...

`-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005`

.. Gradle should use these:

`-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5005`

I am using Gradle 4.10.2.

**Comment by marcphilipp** [comment]
@ChristianCiach Thanks for reporting the issue! Would you be interested in providing a fix?

**Comment by drujd** [comment]
@marcphilipp Would you be interested in merging the PR? This is a real problem.

**Comment by big-guy** [comment]
@drujd I replied on the PR, but I don't think we can accept the PR as it is.

As a workaround, I think your SSH tunnel is a good idea, but I think you could also do something like:
```
tasks.withType(Test) {
    doFirst {
        if (debug) {
            debug = false
            maxParallelForks = 1
            forkOptions.jvmArgs("-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5005")
        }
    }
}
```

My main concern with the PR is that the JDK disabled that behavior for a good reason and we shouldn't go around it without exploring other options.

**Comment by kdubb** [comment]
@big-guy Curious, what is the "good reason" it was changed and where did you find this information?

Also, would allowing a simple property override like "org.gradle.test.debug.address" so we can easily do it from the command line be an agreeable solution?

**Comment by michalsvec** [comment]
org.gradle.test.debug.address would be great!

**Comment by neelkamath** [comment]
It's been 2 years. This issue causes people using JVM 9+ to waste hours/days the first time they run into this (see #13118). I don't understand why we don't just use Christian's PR.

**Comment by maxandersen** [comment]
> tasks.withType(Test) {
>     doFirst {

this workaround fails with the following on Gradle 6.7:

"* What went wrong:
Execution failed for task ':test'.
> The value for property 'enabled' is final and cannot be changed any further."

any other good suggestions ?

**Comment by neelkamath** [comment]
@maxandersen did you see my comment's linked issue?

**Comment by maxandersen** [comment]
@neelkamath I tried, but got different but similar error. In the end I went back to java 8 for now :/

**Comment by oridool** [comment]
So, is there any solution or workaround ?
I'm using gradle 6.8.1 and the workaround above is indeed not working

**Comment by MeFisto94** [comment]
I've also been hit by this issue.
I've then just removed the if debug checks and used `jvmArgs` instead of `forkOptions.jvmArgs` as a quick workaround.

Unfortunately I can't find much more info than http://mail.openjdk.java.net/pipermail/serviceability-dev/2014-May/014828.html, but nothing about the reasoning behind that change.

**Comment by jeremysmitherman** [comment]
If the PR from 2018 is closed with the comment saying we can use workarounds, can we get one for using bootRun instead of tests?

**Comment by maxandersen** [comment]
> Unfortunately I can't find much more info than http://mail.openjdk.java.net/pipermail/serviceability-dev/2014-May/014828.html, but nothing about the reasoning behind that change.

the change was made to avoid you exposing your machine to everyone on your network by default when debugging. its quite reasonable.

**Comment by jeremysmitherman** [comment]
> 
> 
> > Unfortunately I can't find much more info than http://mail.openjdk.java.net/pipermail/serviceability-dev/2014-May/014828.html, but nothing about the reasoning behind that change.
> 
> the change was made to avoid you exposing your machine to everyone on your network by default when debugging. its quite reasonable.

That change is reasonable on the JDK end, the problem is we have to kinda adopt a weird configuration on the gradle end now to make working in a VM/container work because the expected way to configure the networking for the debugger doesn't work.

**Comment by SchwingSK** [comment]
@maxandersen I totally agree with @jeremysmitherman, I just lost a bunch of hours because of that. Thing is, when I need to debug a failing test on a VMs/docker (typical use case: a test fails in a specific environment on our CI), I like the simplicity of:
`gradlew test --tests MyClassTest --debug-jvm`
Now when using a Java > 9, this doesn't work (without any visible reason to the unsuspecting developer).
The 'workaround' is to change the Gradle script for the project each time we want to debug. I don't think that's reasonable.

Moreover, while I see the reasoning behind the Java change (only open the debug port to specific ip besides localhost, because we don't know if the environment is safe in general), it seems to me that the environments running Gradle are not the general case, and end up being either a developper's PC (in which case localhost is fine), and CI servers (which are internal/part of a VPN), so opening connections to these servers are done in a trusted environment. Moreover the ports are only open when the --debug-jvm option is added, it's not like if the ports were opened for each build. But maybe some uses cases differ.

At the very least, when a Java > 9 is detected, a warning should be issued that only localhost connections are accepted. This would save a lot of time. But this doesn't solve the problem, which is to be able to easily debug a test on a distant machine.
I see the following solutions:

- make *:8005 the default if Java >9. The most simple solution for the user, but apparently this isn't acceptable
- add a new option --debug-jvm-address so that we can set it to *:8005. However we lose the simplicity of just using '--debug-jvm'
- add a new option --debug-jvm-unsafe, that automatically sets the address to *:8005. I don't really like the 'unsafe', but I couldn't come up with a better term at the moment
- add a property debugJvmAddress so that it can be specified on the command line via -PdebugJvmAddress=*:5005, or set in gradle.properties

What do you think?

**Comment by nucatus** [comment]
Another workaround is to use GRADLE_OPTS='-Xdebug -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005'

**Comment by eric-gagnon** [comment]
> Another workaround is to use `GRADLE_OPTS='-Xdebug -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005'`

Not sure about that one.

I changed the port number and suspend and I see two initialisation. I suspect that the first connect is actually usefull to debug gradle,  not debug the app started with bootRun

------------

Listening for transport dt_socket at address: 5006

Welcome to Gradle 6.8.3!

...

> Task :bootRun
Listening for transport dt_socket at address: 5005
------------

Maybe there could be a workaround in my particular setup?

On windows, I have the gradle project checked out and I start it like this (i'll eventually mount user folder to be able to use dependency cache) but it's working nicely, only nissing debug.

> winpty docker run -p 7979:7979 -p 5006:5006  -it --rm -u gradle \
> -e GRADLE_OPTS='-Xdebug -agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:5006' \
> -e TZ="America/Montreal" \
> -e LANG="fr_CA.UTF-8" \
> -e LANGUAGE="fr_CA.UTF-8" \
> -e LC_ALL="fr_CA.UTF-8" \
> --mount type=bind,source=$BASE_PROJET_PATH,target=/home/gradle/  \
> gradle:jdk15 gradle bootRun --debug-jvm   \

**Comment by SchwingSK** [comment]
@nucatus this doesn't work, as GRADLE_OPTS are options for the JVM running Gradle, not the forked JVMs running the tests.

**Comment by marcmesh** [comment]
I workarounded this by settings this in my build.gradle:

```
test {
    useJUnitPlatform()

    // If this Property is not set explicitly, the tests are run normally.
    if (System.getProperty('REMOTEDEBUG', 'false') == 'true') {
        jvmArgs '-Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=*:5005'
    }
}
```
and starting my tests with `gradle test -DREMOTEDEBUG=true`. At least it works....

**Comment by BjoernDomschke** [comment]
so does it mean we will stay with the intermediate solution jvmArgs? 
nobody addressing the  DSL 
> test {
   debugOptions {
       enabled = true
      **port = 4455**
        server = true
        suspend = true
    }
}

somebody would be able to specify the `address` attribute - instead just the port

**Comment by ryl** [comment]
I just came across this. Having address in debugOptions would be nice.

**Comment by magneticB** [comment]
Thankfully @marcmesh workaround solves the issue - glad I found this thread.  I was going nuts trying to connect a remote debuger a Docker container running Gradle tests.  Would be really great to get this fixed properly!

**Comment by maxandersen** [comment]
Just got reminded of this and saw that `https://docs.gradle.org/current/userguide/java_testing.html#sec:debugging_java_tests` now mention `host` is possible to specify:

```
test {
    debugOptions {
        enabled = true
        host = 'localhost'
        port = 4455
        server = true
        suspend = true
    }
}
```
 
haven't seen mention of this being possible to specify from gradle command line but at least it is feasible now.

**Comment by magneticB** [comment]
Perhaps the default host value should be * (listen on all interfaces) rather than localhost when running --debug-jvm ?

**Comment by maxandersen** [comment]
I would recommend against it as it exposes your local host debug sessions. 

Should be a flag possible to set via properties somehow to enable for those special cases of running in a container or other "remote" location.

**Comment by alois-git** [comment]
Any update on this ?

**Comment by big-guy** [comment]
7.6 and newer have an option to support this again.

https://docs.gradle.org/7.6/release-notes.html#introduced-support-for-java-9+-network-debugging
